cookbooks/prescribed-velocity/prescribed_velocity.cc

```diff
@@ -0,0 +1,204 @@
+             cell != simulator_access.get_dof_handler().end();
+             ++cell)
+          if (! cell->is_artificial())
+            {
+              fe_values.reinit (cell);
+              std::vector<unsigned int> local_dof_indices(simulator_access.get_fe().dofs_per_cell);
+              cell->get_dof_indices (local_dof_indices);
+
+              for (unsigned int q=0; q<quadrature.size(); q++)
+                // If it's okay to constrain this DOF
+                if (current_constraints.can_store_line(local_dof_indices[q]) &&
+                    !current_constraints.is_constrained(local_dof_indices[q]))
+                  {
+                    const unsigned int component_index = simulator_access.get_fe().system_to_component_index(q).first;
+
+                    // If we're on one of the velocity DOFs
+                    if (component_index >= simulator_access.introspection().component_indices.velocities[0] &&
+                        component_index <= simulator_access.introspection().component_indices.velocities[dim-1])
+                      {
+                        // Which velocity component is this DOF associated with?
+                        const unsigned int component_direction = component_index
+                                                                 - simulator_access.introspection().component_indices.velocities[0];
+
+                        // we get time passed as seconds (always) but may want
+                        // to reinterpret it in years
+                        double time = simulator_access.get_time();
+                        if (simulator_access.convert_output_to_years())
+                          time /= year_in_seconds;
+
+                        prescribed_velocity_indicator_function_2d.set_time (time);
+                        prescribed_velocity_indicator_function_3d.set_time (time);
+                        prescribed_velocity_function_2d.set_time (time);
+                        prescribed_velocity_function_3d.set_time (time);
+
+                        const Point<dim> p = fe_values.quadrature_point(q);
+
+                        double indicator, u_i;
+                        if (dim == 2)
+                          {
+                            indicator = prescribed_velocity_indicator_function_2d.value (reinterpret_cast<const Point<2>&>(p), component_direction);
+                            u_i = prescribed_velocity_function_2d.value (reinterpret_cast<const Point<2>&>(p), component_direction);
+                          }
+                        else
+                          {
+                            indicator = prescribed_velocity_indicator_function_3d.value (reinterpret_cast<const Point<3>&>(p), component_direction);
+                            u_i = prescribed_velocity_function_3d.value (reinterpret_cast<const Point<3>&>(p), component_direction);
+                          }
+
+                        if (indicator > 0)
```

Comment:
if you set values in the input file to zero or one, then maybe comparing against 0.5 would be a better choice.